File: anabeesh/src/main/java/com/cs18/anabeesh/di/application/SchedulersModule.java
package com.cs18.anabeesh.di.application;

import com.cs18.anabeesh.schedulers.ComputationalThread;
import com.cs18.anabeesh.schedulers.ComputationalThreadSchedulers;
import com.cs18.anabeesh.schedulers.IOThread;
import com.cs18.anabeesh.schedulers.IOThreadSchedulers;
import com.cs18.anabeesh.schedulers.MainThread;
import com.cs18.anabeesh.schedulers.MainThreadSchedulers;
import com.cs18.anabeesh.schedulers.ThreadSchedulers;

import dagger.Module;
import dagger.Provides;

/**
 This class is responsible for providing the requested objects for {@link ThreadSchedulers} objects
 */

@Module
public class SchedulersModule {

    @ApplicationScope
    @Provides
    @MainThread
    ThreadSchedulers providesMainThreadSchedulers() {
        return new MainThreadSchedulers();
    }

    @ApplicationScope
    @Provides
    @IOThread
    ThreadSchedulers providesIOThreadSchedulers() {
        return new IOThreadSchedulers();
    }

    @ApplicationScope
    @Provides
    @ComputationalThread
    ThreadSchedulers providesComputationalThreadSchedulers() {
        return new ComputationalThreadSchedulers();
    }
}

File: anabeesh/src/main/java/com/cs18/anabeesh/di/application/SchedulersModule.kt
package com.cs18.anabeesh.di.application

import com.cs18.anabeesh.schedulers.*
import com.cs18.anabeesh.schedulers.qualifier.ComputationalThread
import com.cs18.anabeesh.schedulers.qualifier.IOThread
import com.cs18.anabeesh.schedulers.qualifier.MainThread
import com.cs18.anabeesh.schedulers.qualifier.NetworkThread
import dagger.Module
import dagger.Provides

/**
 * This class is responsible for providing the requested objects for [ThreadSchedulers] objects
 */

@Module
class SchedulersModule {

    @ApplicationScope
    @Provides
    @MainThread
    fun providesMainThreadSchedulers(): ThreadSchedulers = MainThreadSchedulers()

    @ApplicationScope
    @Provides
    @IOThread
    fun providesIOThreadSchedulers(): ThreadSchedulers = IOThreadSchedulers()

    @ApplicationScope
    @Provides
    @ComputationalThread
    fun providesComputationalThreadSchedulers(): ThreadSchedulers = ComputationalThreadSchedulers()

    @ApplicationScope
    @Provides
    @NetworkThread
    fun providesNetworkThreadSchedulers(): ThreadSchedulers = NetworkThreadSchedulers()
}